### DISCUSSION TITLE
ANSISYSRC / DECRC reset color attributes

### DISCUSSION THREAD

**Discussion by alabuzhev** [microsoft/terminal#14699] [General]
### Windows Terminal version

WT 1.16.3463.0, OpenConsole 1.16.2212.12003

### Windows build number

10.0.19045.2364

### Steps to reproduce

https://learn.microsoft.com/en-us/windows/console/console-virtual-terminal-sequences#simple-cursor-positioning

Sequence | Shorthand | Behavior
-- | -- | --
ESC 7 | DECSC | Save Cursor Position in Memory**
ESC 8 | DECSR | Restore Cursor Position from Memory**

https://learn.microsoft.com/en-us/windows/console/console-virtual-terminal-sequences#cursor-positioning

<html>
<body>

Sequence | Code | Description | Behavior
-- | -- | -- | --
ESC [ s | ANSISYSSC | Save Cursor – Ansi.sys emulation | **With no parameters, performs a save cursor operation like DECSC
ESC [ u | ANSISYSRC | Restore Cursor – Ansi.sys emulation | **With no parameters, performs a restore cursor operation like DECRC

Unpack and run the cmd file in WT and OpenConsole:
[test_cusror_position.zip](https://github.com/microsoft/terminal/files/10316390/test_cusror_position.zip)

It sets color attributes and draws a few lines, saving and restoring cursor position along the way.

### Expected Behavior

Since DECRC & ANSISYSRC restore cursor position *only* (according to the documentation above), I expect the color attributes to be applied to *all the printed text*.

It works as expected in conhost.exe 10.0.19041.1566 that comes with the OS:

![image](https://user-images.githubusercontent.com/11453922/209871450-eed9b6bd-b854-464b-b6d1-b753fb472f1f.png)

### Actual Behavior

Color attributes are reset.

OpenConsole 1.16.2212.12003:
![image](https://user-images.githubusercontent.com/11453922/209871612-339ade2d-52f9-4da8-ae08-2f4ac95d4f80.png)

WT 1.16.3463.0:
![image](https://user-images.githubusercontent.com/11453922/209871951-04168e52-2155-4ee9-ba0c-ea7976625843.png)

**Comment by j4james** [comment]
Honestly our documentation on VT sequences is not great. You should really refer to the DEC manuals for any DEC sequences, or the XTerm documentation for most of the non-standard stuff.

And quoting the DEC documentation, `DECSC` is meant to work as follows:

> Saves the following items in the terminal’s memory:
> • Cursor position
> • Character attributes set by the SGR command
> • Character sets (G0, G1, G2, or G3) currently in GL and GR
> • Wrap flag (autowrap or no autowrap)
> • State of origin mode (DECOM)
> • Selective erase attribute
> • Any single shift 2 (SS2) or single shift 3 (SS3) functions sent

And most modern terminals essentially treat the `ANSISC` and `ANSIRC`sequences as aliases for `DECSC` and `DECRC` (XTerm certainly does), so that's why we do that.

I would love one day for us to have a compatibility mode for something like ANSI.SYS and/or ANSI-BBS, but that's not likely to happen anytime soon, and I suspect that's not really what you want anyway.

And the reason it "works" for you in conhost is because that was originally a bug that got fixed a few years back, and it often take years for OpenConsole fixes to make it into the inbox conhost. So I suspect you're only likely to see the correct behavior in a Windows 11 release.

**Comment by alabuzhev** [comment]
> Honestly our documentation on VT sequences is not great

I see, thanks James. It's a bit strange that VT doesn't have a way to save & restore cursor position only, but... it's not the only strange thing about VT.

If it's a documentation issue, it would be nice to fix it and mention somewhere at that page that it is subject to change as VT support improves and one should check other sources to ensure compatibility - it can be frustrating to implement something according to the documentation, see it working in your dev environment and then discover that it only works due to a bug and will be broken in future releases.

**Comment by j4james** [comment]
> If it's a documentation issue, it would be nice to fix it and mention somewhere at that page that it is subject to change as VT support improves and one should check other sources to ensure compatibility

I think that's a good idea. There's an existing issue in the documentation repo for the VT sequences page from 2019 (see https://github.com/MicrosoftDocs/Console-Docs/issues/91), and that doesn't seem likely to be addressed any time soon. But a small update that just directs people to other sources would be a significant improvement in my opinion, and that's the sort of PR that could be reasonably easy for a volunteer contributor to do (although it's probably best to wait to see what the core devs have to say on the subject).
